### PR TITLE
[Proposal] Refactor add_suseconnect_product with retries

### PR DESCRIPTION
**Pull request by jlausuch**
This method is used in multiple places, and we have observed that
it fails randomly due to network issues (timeouts, SCC not reachable,
etc). In most cases, restarting the test helps (unless there is a
major outage that tages long to be fixed).

This code does a similar thing what script_retry does, but with a
bit of more logic and info messages.

- Related ticket: https://progress.opensuse.org/issues/107062

Before:
We do 1 time `SUSEConnect -p` and ONLY retry 1 more time if PH.

After:
We do X retries (default 3) for ANY product.
And keep the same fail message if BETA=1 and PH at the end.

### CODE REVIEW

**Review by foursixnine — APPROVED** [review]
I'm unsure if waiting 60 seconds will be enough or waiting a bit longer would be a better option (say 120s), but overall waiting makes much more sense than constantly retrying specially when the network is loaded.

**Comment by jlausuch** [comment]
> I'm unsure if waiting 60 seconds will be enough or waiting a bit longer would be a better option (say 120s), but overall waiting makes much more sense than constantly retrying specially when the network is loaded.

Yes, maybe 1min is too short. I would even go for 5 min wait :D
Let's see what others think.

**Comment by mloviska** [comment]
> > I'm unsure if waiting 60 seconds will be enough or waiting a bit longer would be a better option (say 120s), but overall waiting makes much more sense than constantly retrying specially when the network is loaded.
> 
> Yes, maybe 1min is too short. I would even go for 5 min wait :D Let's see what others think.

Or increase the value each iteration.

**Comment by jlausuch** [comment]
Some early VRs: (search for `SCC product`)
https://openqa.suse.de/tests/8190066
https://openqa.suse.de/tests/8190065
https://openqa.suse.de/tests/8190011

File: lib/registration.pm [hunk 1]
```diff
@@ -174,25 +174,31 @@ sub add_suseconnect_product {
...
+            sleep 60;
         }
-
-        my $fail_message = ($name =~ /PackageHub/ && check_var('BETA', '1')) ? "PackageHub installation might fail in early development" : undef;
-        assert_script_run("SUSEConnect -p $name/$version/$arch $params", $timeout, $fail_message);
-
-    } elsif ($result && !$retry) {
-        die "SUSEConnect failed activating module $name with exit code (see log output): $result.";
+        else {
+            return 1;
```
Comment (foursixnine) [inline lib/registration.pm:195]: I´d add an info window here (you know, thinking about the reviewers) that there has been too many trials, so aborting.

Comment (pdostal) [reply lib/registration.pm:195]: Isn't the line 201 enough?

**Review by pdostal — APPROVED** [review]

**Review by ilausuch — APPROVED** [review]

File: lib/registration.pm (hunk 1 above)
Comment (jlausuch) [reply lib/registration.pm:195]: No, no, but this line is to return if the command succeded.. something like returning 0 in bash, but in perl we need to return "true" if ok. This condition will hit when `$@` is `undef`, which means that there are no errors from the `eval` section.

Comment (foursixnine) [reply lib/registration.pm:195]: @jlausuch exactly, I need more waay more coffee today xD

**Comment by jlausuch** [comment]
> > > I'm unsure if waiting 60 seconds will be enough or waiting a bit longer would be a better option (say 120s), but overall waiting makes much more sense than constantly retrying specially when the network is loaded.
> > 
> > 
> > Yes, maybe 1min is too short. I would even go for 5 min wait :D Let's see what others think.
> 
> Or increase the value each iteration.

not bad idea :)
iteration_num * 60? e.g. 1) 60  2) 120  3) 180 ...

**Review by tjyrinki — APPROVED** [review]
If we can't get SCC or network to be 100% fault-free, this would be absolutely a great improvement.

**Comment by jlausuch** [comment]
> If we can't get SCC or network to be 100% fault-free, this would be absolutely a great improvement.

Thanks! :)
I still have some hopes we will reach that fault-free level... but for now, we need to settle for some workaround.

**Comment by jlausuch** [comment]
Shall we force default the `$timeout` in this method to e.g. 300s instead of the default by `assert_script_run` (90s)?

**Comment by asmorodskyi** [comment]
I think enough people approve it already so I can merge

**Comment by okurz** [comment]
> > If we can't get SCC or network to be 100% fault-free, this would be absolutely a great improvement.
> 
> Thanks! :) I still have some hopes we will reach that fault-free level... but for now, we need to settle for some workaround.

With this change a lot of motivation to improve will be gone. Did any of you raise the problems to the SCC team? I bet collaborating together can achieve great things by finding out where things might not be 100% fault-free yet. I am sure that a much higher stability can be achieved but only if issues are reported.

**Comment by jlausuch** [comment]
> > > If we can't get SCC or network to be 100% fault-free, this would be absolutely a great improvement.
> > 
> > 
> > Thanks! :) I still have some hopes we will reach that fault-free level... but for now, we need to settle for some workaround.
> 
> With this change a lot of motivation to improve will be gone. Did any of you raise the problems to the SCC team? I bet collaborating together can achieve great things by finding out where things might not be 100% fault-free yet. I am sure that a much higher stability can be achieved but only if issues are reported.

I think this situation hasn't changed for long time... And.. are you sure the problem is in SCC network? It would be multiple places... random slowdowns in our openQA worker infrastructure as well, who knows? if we are to blame SCC we need proof for that.

**Comment by okurz** [comment]
> I think this situation hasn't changed for long time... And.. are you sure the problem is in SCC network? It would be multiple places... random slowdowns in our openQA worker infrastructure as well, who knows? if we are to blame SCC we need proof for that.

I did not suggest to blame anyone :) I am suggesting to collaborate on the investigation so that we can identify which component might be the culprit.

**Comment by jlausuch** [comment]
It seems it doesn't work as expected: https://openqa.suse.de/tests/8252341#step/suseconnect_scc/42

**Comment by jlausuch** [comment]
Followu-up: https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/14402

**Pull request merged**